### PR TITLE
Revert "Revert D22812445: Update TensorPipe submodule"

### PR DESCRIPTION
**Pull request by malfet**
This reverts commit 352e15f1a2df6b93689a28c14b71a0f87f29cc02.

### CODE REVIEW

**Comment by lw** [comment]
I've been looking at the failure reported in #42372: 
```
Jul 31 09:10:34 ImportError: dlopen(/Users/distiller/project/conda/envs/env_py3.8_0_20200731/lib/python3.8/site-packages/torch/_C.cpython-38-darwin.so, 2): Library not loaded: @rpath/libuv.1.dylib
Jul 31 09:10:34   Referenced from: /Users/distiller/project/conda/envs/env_py3.8_0_20200731/lib/python3.8/site-packages/torch/lib/libtensorpipe.dylib
```
I think a relevant hint to that problem can be found earlier on in the logs:
```
Jul 31 07:14:01 CMake Warning (dev) at third_party/tensorpipe/cmake/Options.cmake:16 (option):
Jul 31 07:14:01   Policy CMP0077 is not set: option() honors normal variables.  Run "cmake
Jul 31 07:14:01   --help-policy CMP0077" for policy details.  Use the cmake_policy command to
Jul 31 07:14:01   set the policy and suppress this warning.
Jul 31 07:14:01 
Jul 31 07:14:01   For compatibility with older versions of CMake, option is clearing the
Jul 31 07:14:01   normal variable 'TP_BUILD_LIBUV'.
Jul 31 07:14:01 Call Stack (most recent call first):
Jul 31 07:14:01   third_party/tensorpipe/CMakeLists.txt:19 (include)
Jul 31 07:14:01 This warning is for project developers.  Use -Wno-dev to suppress it.
```
So I suspect what's happening is that in the PyTorch CMake file we're setting `set(TP_BUILD_LIBUV ON)` (without any `CACHE BOOL "" FORCE` arguments) and then in TensorPipe's CMake we're doing `option(TP_BUILD_LIBUV "..." OFF)` which _overrides_ that option, meaning that we don't force TensorPipe to build libuv from its own submodule.

This doesn't cause problems on Linux because there is no system libuv available and thus TensorPipe ends up building its submodule anyways. But with MacOS we do install a system libuv, in order for Gloo to pick it up, and then TensorPipe does so too.

Question: should TensorPipe use the system libuv under MacOS, or still build its own one (and link to it statically)?

**Comment by lw** [comment]
As for the failure that happened here: https://app.circleci.com/pipelines/github/pytorch/pytorch/195896/workflows/45ddda15-6ac4-4b3c-9683-f26de7f516fe/jobs/6474787

The failure is caused by a newer revision of PyTorch using an old revision of TensorPipe as a submodule. At first I thought it was an issue with the CI but in fact it's a different problem. The CI run was on branch `gh/iseeyuan/80/head`, commit a723a43f2f647ce33b7c203ac01b8c650bf69876. The history before that commit is [here](https://github.com/pytorch/pytorch/commits/a723a43f2f647ce33b7c203ac01b8c650bf69876). And we can see that in [one of the commits in that branch](https://github.com/pytorch/pytorch/commit/34bc7fd2e3787d9bf50b6a1f712a20974feb51c3) the submodule update of TensorPipe was reverted without the other changes to the PyTorch repo being reverted as well.

So, in short, this was a "user error". It probably wasn't intentional and may in fact have been caused by ghstack&co., which in particular aren't able to manage submodule updates well (AFAIK) and thus may have messed up. I don't know if this is worth further investigation...
